libheif/codecs/uncompressed_box.cc

```diff
@@ -394,68 +391,164 @@ Error Box_cmpC::write(StreamWriter& writer) const
+      unitInfo.unit_offset = implied_offset;
+    } else if (unit_offset_code == 1) {
+      unitInfo.unit_offset = range.read16();
+    } else if (unit_offset_code == 2) {
+      unitInfo.unit_offset = range.read24();
+    } else if (unit_offset_code == 3) {
+      unitInfo.unit_offset = range.read32();
+    } else if (unit_offset_code == 4) {
+      unitInfo.unit_offset = range.read64();
+    } else {
+      return Error(heif_error_Usage_error, heif_suberror_Unsupported_parameter, "Unsupported icef unit offset code");
+    }
+    if (unit_size_code == 0) {
+      unitInfo.unit_size = range.read8();
+    } else if (unit_size_code == 1) {
+      unitInfo.unit_size = range.read16();
+    } else if (unit_size_code == 2) {
+      unitInfo.unit_size = range.read24();
+    } else if (unit_size_code == 3) {
+      unitInfo.unit_size = range.read32();
+    } else if (unit_size_code == 4) {
+      unitInfo.unit_size = range.read64();
     } else {
-      return Error(heif_error_Usage_error, heif_suberror_Unsupported_parameter, "Unsupported icbr version");
+      return Error(heif_error_Usage_error, heif_suberror_Unsupported_parameter, "Unsupported icef unit size code");
     }
+    implied_offset += unitInfo.unit_size;
     if (range.get_error() != Error::Ok) {
       return range.get_error();
     }
-    m_ranges.push_back(byteRange);
+    m_unit_infos.push_back(unitInfo);
   }
   return range.get_error();
 }
 
 
-std::string Box_icbr::dump(Indent& indent) const
+std::string Box_icef::dump(Indent& indent) const
 {
   std::ostringstream sstr;
   sstr << Box::dump(indent);
-  sstr << indent << "num_ranges: " << m_ranges.size() << "\n";
-  for (ByteRange range: m_ranges) {
-    sstr << indent << "range_offset: " << range.range_offset << ", range_size: " << range.range_size << "\n";
+  sstr << indent << "num_compressed_units: " << m_unit_infos.size() << "\n";
+  for (CompressedUnitInfo unit_info: m_unit_infos) {
+    sstr << indent << "unit_offset: " << unit_info.unit_offset << ", unit_size: " << unit_info.unit_size << "\n";
   }
   return sstr.str();
 }
 
-Error Box_icbr::write(StreamWriter& writer) const
+Error Box_icef::write(StreamWriter& writer) const
 {
   size_t box_start = reserve_box_header_space(writer);
 
-  writer.write32((uint32_t)m_ranges.size());
-  for (ByteRange range: m_ranges) {
-    if (get_version() == 1) {
-      writer.write64(range.range_offset);
-      writer.write64(range.range_size);
-    } else if (get_version() == 0) {
-      writer.write32((uint32_t)range.range_offset);
-      writer.write32((uint32_t)range.range_size);
+  int unit_offset_nbbits = 16;
+  int unit_size_nbbits = 8;
+  uint64_t implied_offset = 0;
+  bool can_use_implied_offsets = true;
+  for (CompressedUnitInfo unit_info: m_unit_infos) {
+    if (unit_info.unit_offset != implied_offset) {
+      can_use_implied_offsets = false;
+    }
+    implied_offset += unit_info.unit_size;
+    while (unit_info.unit_offset >= (1ULL << unit_offset_nbbits)) {
```

Comment:
clang-tidy says:
````
uncompressed_box.cc:476:43: error: The result of the left shift is undefined due to shifting by '64', which is greater or equal to the width of type 'unsigned long long' [clang-analyzer-core.UndefinedBinaryOperatorResult,-warnings-as-errors]
    while (unit_info.unit_offset >= (1ULL << unit_offset_nbbits))
````

I think this is a false positive, but you may want to check too.

Reply:
Here is the link to the logical reasoning:
https://github.com/strukturag/libheif/actions/runs/10369014420/job/28703823871#step:6:521

Reply:
Now I see why this is wrong. The computation of `unit_offset_nbbits` is in a for-loop and leads to the error when it has reached 64bit in a previous iteration.
Please check the computation. Maybe it has to be outside the for-loop.

Reply:
I have an idea for a cleaner approach.  Will update.